r/src/altrep.cpp

```diff
@@ -43,144 +50,356 @@ extern "C" {
+    Rprintf("arrow::Array<%s, %d nulls, %s> len=%d, Array=<%p>\n",
+            array_->type()->ToString().c_str(), array_->null_count(),
+            IsMaterialized() ? "materialized" : "not materialized", array_->length(),
+            array_.get());
+    inspect_subtree(data1(), pre, deep + 1, pvec);
+    if (IsMaterialized()) {
+      inspect_subtree(data2(), pre, deep + 1, pvec);
+    }
+    return TRUE;
+  }
+
+  const void* Dataptr_or_null() {
+    if (array_->null_count() == 0) {
+      return reinterpret_cast<const void*>(const_array_data());
+    }
 
-  static void Init(R_altrep_class_t class_t, DllInfo* dll) {
-    // altrep
-    R_set_altrep_Length_method(class_t, ArrayNoNull::Length);
-    R_set_altrep_Inspect_method(class_t, ArrayNoNull::Inspect);
-    R_set_altrep_Duplicate_method(class_t, ArrayNoNull::Duplicate);
+    if (IsMaterialized()) {
+      return DATAPTR_RO(data2());
+    }
 
-    // altvec
-    R_set_altvec_Dataptr_method(class_t, ArrayNoNull::Dataptr);
-    R_set_altvec_Dataptr_or_null_method(class_t, ArrayNoNull::Dataptr_or_null);
+    return NULL;
   }
-};
 
-struct DoubleArrayNoNull {
-  static R_altrep_class_t class_t;
+  // force materialization, and then return data ptr
+  void* Dataptr(Rboolean writeable) {
+    if (array_->null_count() == 0) {
+      return reinterpret_cast<void*>(array_data());
+    }
 
-  static void Init(DllInfo* dll) {
-    class_t = R_make_altreal_class("array_nonull_dbl_vector", "arrow", dll);
-    ArrayNoNull<REALSXP>::Init(class_t, dll);
-    R_set_altreal_No_NA_method(class_t, ArrayNoNull<REALSXP>::No_NA);
+    Materialize();
+    return DATAPTR(data2());
   }
 
-  static SEXP Make(const std::shared_ptr<Array>& array) {
-    return ArrayNoNull<REALSXP>::Make(class_t, array);
+  // by design, there are missing values in this vector
+  int No_NA() { return array_->null_count() != 0; }
+
+  // There are no guarantee that the data is the R sentinel,
+  // so we have to treat it specially
+  data_type Elt(R_xlen_t i) {
+    return array_->IsNull(i) ? cpp11::na<data_type>()
+                             : array_->data()->template GetValues<data_type>(1)[i];
   }
-};
 
-struct Int32ArrayNoNull {
-  static R_altrep_class_t class_t;
+  SEXP alt_;
+  const std::shared_ptr<Array>& array_;
 
-  static void Init(DllInfo* dll) {
-    class_t = R_make_altinteger_class("array_nonull_int_vector", "arrow", dll);
-    ArrayNoNull<INTSXP>::Init(class_t, dll);
-    R_set_altinteger_No_NA_method(class_t, ArrayNoNull<INTSXP>::No_NA);
+  const data_type* const_array_data() const {
+    return array_->data()->template GetValues<data_type>(1);
   }
 
-  static SEXP Make(const std::shared_ptr<Array>& array) {
-    return ArrayNoNull<INTSXP>::Make(class_t, array);
+  data_type* array_data() const {
+    return const_cast<data_type*>(array_->data()->template GetValues<data_type>(1));
+  }
+
+  R_xlen_t Get_region(R_xlen_t i, R_xlen_t n, data_type* buf) {
+    const auto& slice = array_->Slice(i, n);
```

Comment:
I don't think you can take a const reference here. `Array::Slice` is returning a new `shared_ptr`.